### PR TITLE
List and ListStream implementation

### PR DESCRIPTION
**Pull request by andrebsguedes**
Exposes list functionality through buffered and streaming APIs

### CODE REVIEW

**Comment by nickrobinson251** [comment]
will try take a look at this tomorrow, but Justin probably better placed to review

one random thought: i wonder if we should maybe just put this all in its own file, since it's pretty separate to the get/put implementation -- would that make the codebase more maintainable? Just a passing thought, so take it or leave it 😊

File: src/lib.rs [hunk 1]
```diff
@@ -633,6 +634,373 @@ pub extern "C" fn delete(
...
+}
+
+impl ListResponseGuard {
+    unsafe fn new(response_ptr: *mut ListResponse, handle: *const c_void) -> ListResponseGuard {
+        let response = unsafe { &mut (*response_ptr) };
+        response.result = CResult::Uninitialized;
+
+        ListResponseGuard { response, handle }
+    }
+    fn success(self, entries: Vec<ObjectMeta>) { // TODO entries
```
Comment (jfunstonRAI) [inline src/lib.rs]: Leftover TODO?

File: src/lib.rs [hunk 2]
```diff
@@ -633,6 +634,373 @@ pub extern "C" fn delete(
...
+    handle: *const c_void
+) -> CResult {
+    let response = unsafe { ListResponseGuard::new(response, handle) };
+    let prefix = unsafe { std::ffi::CStr::from_ptr(prefix) };
+    let prefix: Path = prefix.to_str().expect("invalid utf8").try_into().unwrap();
+    let config = unsafe { & (*config) };
+
+    match RT.get() {
+        Some(runtime) => {
+            runtime.spawn(async move {
```
Comment (jfunstonRAI) [inline src/lib.rs]: Shouldn't we use our `buffer_unordered` dispatch queue?

Comment (andrebsguedes) [reply src/lib.rs]: I think it is better to leave the low throughput operations out of the queue in order to not introduce latency due to a bunch of get or put requests that may be happening in the background. Unless we are expecting hundreds of list operations per second, then it should be going through the queue, what do you think?

File: src/lib.rs (hunk 1 above)
Comment (andrebsguedes) [reply src/lib.rs]: Good catch

File: src/lib.rs (hunk 2 above)
Comment (jfunstonRAI) [reply src/lib.rs]: I don't think we should make assumptions on which operations might be higher throughput or higher priority, it seems better to make it consistent. I know for blobgc for example they will be doing continuous spot checks for unindexed pages using listing.

**Review by jfunstonRAI — APPROVED** [review]

**Pull request merged**